File: android/java/org/chromium/chrome/browser/allow2/util/Allow2PinVerifier.kt
/* Copyright (c) 2024 The Brave Authors. All rights reserved.
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this file,
 * You can obtain one at https://mozilla.org/MPL/2.0/. */

package org.chromium.chrome.browser.allow2.util

import org.chromium.chrome.browser.allow2.models.Allow2Child
import java.security.MessageDigest

/**
 * Utility class for verifying PINs using constant-time comparison.
 * This prevents timing attacks that could leak PIN information.
 */
object Allow2PinVerifier {

    /**
     * Verify a PIN against a child's stored hash and salt.
     *
     * @param pin The PIN entered by the user
     * @param child The child whose PIN is being verified
     * @return true if the PIN is correct, false otherwise
     */
    fun verify(pin: String, child: Allow2Child): Boolean {
        if (!child.hasPIN()) {
            // No PIN set means access granted
            return true
        }

        val computedHash = computeHash(pin, child.pinSalt)
        val storedHash = child.pinHash

        // Handle "sha256:" prefix if present
        val normalizedStored = if (storedHash.startsWith("sha256:")) {
            storedHash.substring(7)
        } else {
            storedHash
        }

        return constantTimeCompare(computedHash, normalizedStored)
    }

    /**
     * Compute SHA-256 hash of PIN + salt.
     *
     * @param pin The PIN to hash
     * @param salt The salt to combine with the PIN
     * @return Hexadecimal string of the hash
     */
    private fun computeHash(pin: String, salt: String): String {
        val combined = pin + salt
        val digest = MessageDigest.getInstance("SHA-256")
        val hashBytes = digest.digest(combined.toByteArray(Charsets.UTF_8))
        return hashBytes.joinToString("") { "%02x".format(it) }
    }

    /**
     * Constant-time string comparison to prevent timing attacks.
     * This ensures the comparison takes the same amount of time regardless
     * of where the strings differ.
     *
     * @param a First string to compare
     * @param b Second string to compare
     * @return true if strings are equal, false otherwise
     */
    private fun constantTimeCompare(a: String, b: String): Boolean {
        if (a.length != b.length) {
            // Still do comparison to maintain constant timing characteristics
            // even when lengths differ
            var result: Int = a.length xor b.length
            val minLen = minOf(a.length, b.length)
            for (i in 0 until minLen) {
                result = result or (a[i].code xor b[i].code)
            }
            return false
        }

        var result = 0
        for (i in a.indices) {
            result = result or (a[i].code xor b[i].code)
        }
        return result == 0
    }

    /**
     * Validate PIN format (4-6 digits).
     *
     * @param pin The PIN to validate
     * @return true if PIN format is valid
     */
    fun isValidPinFormat(pin: String): Boolean {
        return pin.length in 4..6 && pin.all { it.isDigit() }
    }

    /**
     * Mask a PIN for display (show asterisks).
     *
     * @param length Number of characters to show
     * @return String of asterisks
     */
    fun maskPin(length: Int): String {
        return "*".repeat(length)
    }
}

File: ios/brave-ios/Sources/Brave/Frontend/ParentalFreedom/Allow2PINVerifier.swift
// Copyright 2024 The Brave Authors. All rights reserved.
// This Source Code Form is subject to the terms of the Mozilla Public
// License, v. 2.0. If a copy of the MPL was not distributed with this
// file, You can obtain one at https://mozilla.org/MPL/2.0/.

import CryptoKit
import Foundation

// MARK: - Allow2 PIN Verifier

/// Utility for secure PIN verification using constant-time comparison
/// to prevent timing attacks
public enum Allow2PINVerifier {

  // MARK: - Hash Formats

  /// Supported hash formats
  public enum HashFormat: String {
    case sha256 = "sha256"
    case sha512 = "sha512"
  }

  // MARK: - PIN Verification

  /// Verify a PIN against a stored hash
  /// - Parameters:
  ///   - pin: The entered PIN to verify
  ///   - child: The child whose PIN hash to verify against
  /// - Returns: True if PIN is valid
  public static func verifyPIN(_ pin: String, for child: Allow2Child) -> Bool {
    // Hash the entered PIN with the child's salt
    let computedHash = hashPIN(pin, salt: child.pinSalt)

    // Perform constant-time comparison
    return constantTimeCompare(computedHash, child.pinHash)
  }

  /// Hash a PIN with the given salt
  /// - Parameters:
  ///   - pin: PIN to hash
  ///   - salt: Salt value
  ///   - format: Hash format (default: sha256)
  /// - Returns: Formatted hash string (e.g., "sha256:abcd1234...")
  public static func hashPIN(
    _ pin: String,
    salt: String,
    format: HashFormat = .sha256
  ) -> String {
    let saltedPin = pin + salt
    let inputData = Data(saltedPin.utf8)

    switch format {
    case .sha256:
      let hash = SHA256.hash(data: inputData)
      let hashString = hash.compactMap { String(format: "%02x", $0) }.joined()
      return "\(format.rawValue):\(hashString)"

    case .sha512:
      let hash = SHA512.hash(data: inputData)
      let hashString = hash.compactMap { String(format: "%02x", $0) }.joined()
      return "\(format.rawValue):\(hashString)"
    }
  }

  // MARK: - Constant-Time Comparison

  /// Perform a constant-time comparison of two strings
  /// This prevents timing attacks by ensuring the comparison takes
  /// the same amount of time regardless of where differences occur
  /// - Parameters:
  ///   - a: First string
  ///   - b: Second string
  /// - Returns: True if strings are equal
  public static func constantTimeCompare(_ a: String, _ b: String) -> Bool {
    // First, ensure same length (this is not constant-time, but the
    // hash lengths should always match for valid comparisons)
    guard a.count == b.count else {
      return false
    }

    var result: UInt8 = 0

    // XOR each byte pair
    for (byteA, byteB) in zip(a.utf8, b.utf8) {
      result |= byteA ^ byteB
    }

    return result == 0
  }

  /// Perform constant-time comparison of two byte arrays
  /// - Parameters:
  ///   - a: First byte array
  ///   - b: Second byte array
  /// - Returns: True if arrays are equal
  public static func constantTimeCompare(_ a: [UInt8], _ b: [UInt8]) -> Bool {
    guard a.count == b.count else {
      return false
    }

    var result: UInt8 = 0

    for i in 0..<a.count {
      result |= a[i] ^ b[i]
    }

    return result == 0
  }

  /// Perform constant-time comparison of two Data objects
  /// - Parameters:
  ///   - a: First Data
  ///   - b: Second Data
  /// - Returns: True if Data objects are equal
  public static func constantTimeCompare(_ a: Data, _ b: Data) -> Bool {
    guard a.count == b.count else {
      return false
    }

    var result: UInt8 = 0

    for i in 0..<a.count {
      result |= a[i] ^ b[i]
    }

    return result == 0
  }

  // MARK: - PIN Validation

  /// Validate PIN format
  /// - Parameter pin: PIN to validate
  /// - Returns: True if PIN format is valid
  public static func isValidPINFormat(_ pin: String) -> Bool {
    // PIN must be exactly 4 digits
    guard pin.count == 4 else {
      return false
    }

    // All characters must be digits
    return pin.allSatisfy { $0.isNumber }
  }

  /// Validate PIN strength (optional - not used for simple 4-digit PINs)
  /// - Parameter pin: PIN to check
  /// - Returns: Strength assessment
  public static func pinStrength(_ pin: String) -> PINStrength {
    guard isValidPINFormat(pin) else {
      return .invalid
    }

    // Check for repeating digits (e.g., 1111)
    if Set(pin).count == 1 {
      return .weak
    }

    // Check for sequential digits (e.g., 1234, 4321)
    let digits = pin.compactMap { Int(String($0)) }
    let isAscending = zip(digits, digits.dropFirst()).allSatisfy { $1 == $0 + 1 }
    let isDescending = zip(digits, digits.dropFirst()).allSatisfy { $1 == $0 - 1 }

    if isAscending || isDescending {
      return .weak
    }

    // Check for common PINs
    let commonPINs = ["0000", "1111", "2222", "3333", "4444", "5555",
                      "6666", "7777", "8888", "9999", "1234", "4321",
                      "0123", "3210", "1122", "2211", "1212", "2121"]

    if commonPINs.contains(pin) {
      return .weak
    }

    return .strong
  }

  // MARK: - Hash Parsing

  /// Parse hash format from a stored hash string
  /// - Parameter hashString: Stored hash (e.g., "sha256:abcd1234...")
  /// - Returns: Tuple of (format, hashValue) or nil if invalid
  public static func parseHash(_ hashString: String) -> (format: HashFormat, hash: String)? {
    let components = hashString.split(separator: ":", maxSplits: 1)
    guard components.count == 2 else {
      return nil
    }

    let formatString = String(components[0])
    let hashValue = String(components[1])

    guard let format = HashFormat(rawValue: formatString) else {
      return nil
    }

    return (format, hashValue)
  }
}

// MARK: - PIN Strength

public enum PINStrength {
  case invalid
  case weak
  case strong

  public var description: String {
    switch self {
    case .invalid:
      return "Invalid PIN format"
    case .weak:
      return "Weak PIN - easily guessable"
    case .strong:
      return "Strong PIN"
    }
  }

  public var isAcceptable: Bool {
    return self == .weak || self == .strong
  }
}

// MARK: - String Extension

extension String {
  /// Performs a constant-time comparison to prevent timing attacks
  /// - Parameter other: String to compare against
  /// - Returns: True if strings are equal
  func secureCompare(to other: String) -> Bool {
    return Allow2PINVerifier.constantTimeCompare(self, other)
  }
}
